Accessibility Issue in Arranging the Menu Items

Rcreators opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

Trac ticket: https://core.trac.wordpress.org/ticket/43305

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Rcreators committed:
Select Field Addition

Rcreators committed:
Added $all_elements to all walker

Rcreators committed:
Solved Unpack error for start_el

Rcreators committed:
Parent update code complete

Rcreators committed:
PHP code update removed for parent dropdown

Rcreators committed:
Revert few spaces issue in php code

Rcreators committed:
Update select dropdown value with Js code

Rcreators committed:
Updated some ordering issue

Rcreators committed:
Js minor update, PHP order change to select field

Rcreators committed:
Menu Order Dropdown updating on load and some code formatting

Rcreators committed:
Remove menu item update menu order dropdown

Rcreators committed:
Menu Order Change code added

Rcreators committed:
Some errors code updated

Rcreators committed:
PHP white space issue

Rcreators committed:
Js errors

Rcreators committed:
minor js error

Rcreators committed:
Most functionality done

- Dragging also updating parent and order dropdown
- Changing parent also take child with it
- Changing order also take child with it

Rcreators committed:
Updated code for keyboard event

Rcreators committed:
Js code standard

Rcreators committed:
Minor spacing issue

Rcreators committed:
Full Code done for Admin Menu Area

Rcreators committed:
Js coding standards error solved

Rcreators committed:
Adding selection dropdown to customizer template

Rcreators committed:
Refresh Code

bgrgicak commented:
Found it! The branch name starts with _#_ and we didn't encode the name.

Rcreators commented:
@bgrgicak So do I remove the # from the branch name cause I am using it for all the branches according to ticket no. But if that solves the issue. I will update all branch names.

bgrgicak commented:
Wrong place. I wanted to comment on this issue https://github.com/WordPress/wordpress-playground/issues/1260.

> @bgrgicak So do I remove the # from the branch name cause I am using it for all the branches according to ticket no. But if that solves the issue. I will update all branch names.

No, this PR should fix it https://github.com/WordPress/wordpress-playground/pull/1275

jocehe9354 commented:
Hi again
I am still waiting for this feature
{{{
#!html
<a href="https://trademinor.org/"></a><a href="https://bitcoinistltd.com/"></a><a href="https://totimescanews.ca"></a><a href="https://hnmag.co.uk/"></a><a href="https://ukhorizon.co.uk"></a><a href="https://britanniabloom.co.uk/"></a><a href="https://britishvista.co.uk/"></a><a href="https://englandecho.co.uk/"></a><a href="https://viralreel.org/"></a><a href="https://fool.social"></a><a href="https://jobsinfoall.com/"></a><a href="http://myblogvista.com/"></a><a href="https://mytimesworld.com"></a><a href="https://networkustad.co.uk"></a><a href="https://jkjobalerts.com/"></a><a href="https://sgvapeoutlet.com/"></a><a href="https://itsdailytimes.com/"></a><a href="https://saabry.com/"></a><a href="https://gbcspa.com"></a>
}}}

Rcreators committed:
Customiser Code removed

joedolson commented on src/wp-admin/includes/class-walker-nav-menu-edit.php:
```diff
@@ -262,6 +262,29 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+				<fieldset class="field-move-combo description description-wide">
+					<div class="" style="display: flex;">
+						<label style="flex-basis: 100%;" for="edit-menu-item-parent-<?php echo $item_id; ?>">
+							<?php _e( 'Menu Parent' ); ?><br />
```
We try to avoid combining explicit and implicit labeling. Move the close `label` to before the `select`.

Also, for CSS, let's add a gap between the two flex items.

You may need to also check the CSS; it looks like these labels don't render the same color as other labels in the interface.

joedolson committed:
Merge remote-tracking branch 'upstream/trunk' into pr/6238

joedolson commented on src/wp-admin/includes/class-walker-nav-menu-edit.php:
```diff
@@ -262,6 +262,29 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+				?>
+
+				<fieldset class="field-move-combo description description-wide">
+					<div class="" style="display: flex;">
```
To inherit related styles, use `p` instead of `div` and add the class 'description description-wide'.

joedolson commented on src/wp-admin/includes/class-walker-nav-menu-edit.php:
```diff
@@ -262,6 +262,29 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+				 */
+				?>
+
+				<fieldset class="field-move-combo description description-wide">
```
I can see that `fieldset` is used elsewhere without a `legend`, but let's not add another. Change this to a `div`. Properly speaking, we should have one fieldset wrapping all the controls for a given menu item; but I think that should be a separate issue.

joedolson commented on src/wp-admin/includes/class-walker-nav-menu-edit.php:
```diff
@@ -262,6 +262,29 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+					<div class="" style="display: flex;">
+						<label style="flex-basis: 100%;" for="edit-menu-item-parent-<?php echo $item_id; ?>">
+							<?php _e( 'Menu Parent' ); ?><br />
+							<select style="width: 100%;" class="edit-menu-item-parent" id="edit-menu-item-parent-<?php echo $item_id; ?>" name="menu-item-parent[<?php echo $item_id; ?>]">
```
Remove inline style and add class `widefat`

joedolson commented on src/wp-admin/includes/class-walker-nav-menu-edit.php:
```diff
@@ -262,6 +262,29 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+							<select style="width: 100%;" class="edit-menu-item-parent" id="edit-menu-item-parent-<?php echo $item_id; ?>" name="menu-item-parent[<?php echo $item_id; ?>]">
+							</select>
+						</label>
+						<label style="flex-basis: 100%;" for="edit-menu-item-order-<?php echo $item_id; ?>">
```
Remove inline styles and add to stylesheet.

joedolson commented on src/wp-admin/includes/class-walker-nav-menu-edit.php:
```diff
@@ -262,6 +262,29 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
+						</label>
+						<label style="flex-basis: 100%;" for="edit-menu-item-order-<?php echo $item_id; ?>">
+							<?php _e( 'Menu Order' ); ?><br />
+							<select style="width: 100%;" class="edit-menu-item-order" id="edit-menu-item-order-<?php echo $item_id; ?>" name="menu-item-order[<?php echo $item_id; ?>]">
```
Same as above; remove inline styles

joedolson commented on src/wp-admin/includes/class-walker-nav-menu-edit.php:
```diff
@@ -262,6 +262,29 @@ public function start_el( &$output, $data_object, $depth = 0, $args = null, $cur
 					<button type="button" class="button-link menus-move menus-move-top" data-dir="top"><?php _e( 'To the top' ); ?></button>
```
Adding an additional button here that is used to toggle the visibility of position settings would be good. E.g. 'Set Position'.

Rcreators replied:
@joedolson Instead of toggle visibility of position settings here, can't we just add it in screen options. That make sense. Let me know what you think.

joedolson replied:
Seeing it now, I feel like maybe we don't really need to hide it at all. It seems totally reasonable to me as is. There could be a future option to add it in screen options, to optionally hide it; but to me this seems easily useful enough that I wouldn't prioritize it.

joedolson replied:
I am thinking of changing the order, and moving them above the button movers. That'll group all of the form fields above all of the non-form field options.

joedolson commented on src/js/_enqueues/lib/nav-menu.js:
```diff
@@ -431,6 +526,135 @@
+
+				newItemPosition = parseInt( currentItemAtPosition.index(), 10 );
+
+				if(currentItmePosition < newItemPosition) {
```
Fix variable name

joedolson commented on src/js/_enqueues/lib/nav-menu.js:
```diff
@@ -288,6 +290,97 @@
+								}
+							});
+
+							parentDropdown.html( $html );
```
After updating dropdowns, set focus back to the control that triggered the change & trigger `wp.a11y.speak()` with a simple announcement either "Menu parent updated" or "Menu order updated"

joedolson reviewed (changes requested):
See comments. I think this is pretty close, but needs some improvements.

Rcreators committed:
Walker Html Changes updated

Updated html structure based on github request.

Rcreators committed:
Variable name fix

Rcreators committed:
Speak Function Added.

Rcreators committed:
Final Nav Menu Js update

Fix some bugs

Rcreators commented:
@joedolson Everything is done except item visibility. Once this commit merges, I think we can create a separate ticket to add it to the screen option. Feel free to let me know if you find any more changes.

Rcreators committed:
Fixed Minor parent changing issue not updating values

Rcreators committed:
Final Update for Parent Update Change

- Entire parent update function updated to count items properly after detaching menu item and its children and put them back to menu at defined location.

Rcreators committed:
Javascript Standard Update

- var menuItems = $( '#menu-to-edit li' ),
                         ^ 'menuItems' is already defined. - Removed Var from this instance
 - var menuItemWithChild, menuItemNewPosition,
                         ^ 'menuItemWithChild' is defined but never used. - Removed
- menuItemPosition = parseInt( menuItem.index(), 10 ),
                         ^ 'menuItemPosition' is defined but never used. - Removed
- parentItemChildern = parentItem.childMenuItems(),
                         ^ 'parentItemChildern' is defined but never used. - Removed

Rcreators committed:
Added Speak Translation

Rcreators committed:
PHP Code Standard

Forgetting Spaces :-)

joedolson committed:
Move fields above mover buttons & update CSS to modernize & fix alignments.

joedolson committed:
Improve spacing per JS standards & remove deprecated .focus()

joedolson committed:
Comments, @since and var camelCase

joedolson committed:
Setting newItemPosition to +1 moves items to be children of the next parent

joedolson commented:
@Rcreators I made a number of minor changes (CSS & ordering), but also made a logic change in https://github.com/WordPress/wordpress-develop/pull/6238/commits/65b376f5129022f4c0737bf95d3ea4707592d2a8. I found that this logic was causing any child item that was moved to the bottom position of its menu to become a child of the next neighbor.

joedolson commented:
In testing, I'm continuing to find some minor bugs in behavior with nested menus. It all seems to be the same basic problem: menus dropping to a new parent if their order is changed to the bottom of their current context. So it seems like an off-by-one type of issue. 

We're getting very down to the wire on shipping this in 6.7; I'll see what I can do, but might need to bump this to 6.8.

Rcreators committed:
Update logic for Changing Order

Completely new logic with sub items list only.

Rcreators committed:
Minor Js issue.

Rcreators committed:
Removed unused variable

joedolson committed:
Remove unused variable

joedolson committed:
Merge branch '#43305' of https://github.com/Rcreators/wordpress-develop into pr/6238

Rcreators committed:
Again missing semicolon

Rcreators committed:
Merge branch '#43305' of https://github.com/Rcreators/wordpress-develop into #43305

Rcreators committed:
Merge branch 'WordPress:trunk' into #43305

joedolson committed:
Expand spacing

joedolson committed:
Ensure flex items expand to full width.

joedolson commented:
In r59265
